### PR TITLE
Kerynn Davis Mid Mod IC

### PR DESCRIPTION
**Pull request by Kerynn**
I felt pretty comfortable with this IC. I got through the user stories and started the extension. I was able to get the initial studio show page set up (showing the studio attributes), but did not get to having the actors for any of the studio's movies display.

### CODE REVIEW

File: app/controllers/movie_actors_controller.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class MovieActorsController < ApplicationController
+
+  def create 
+    movie = Movie.find(params[:id])
+    actor = Actor.find(params[:actor_id])
+    MovieActor.create!(movie_id: movie.id, actor_id: actor.id)
```
Comment (cjsim89) [inline app/controllers/movie_actors_controller.rb:6]: Nice job creating this 3rd controller, I think it makes the most sense in these situations.

File: app/models/movie.rb [hunk 2]
```diff
@@ -1,3 +1,17 @@
...
+  has_many :movie_actors
+  has_many :actors, through: :movie_actors
+
+  def sort_actors
+    actors.order(:age)
+  end
+
+  def average_age_of_actors
+    if actors.empty?
+      0
```
Comment (cjsim89) [inline app/models/movie.rb:12]: Good AR overall; though I'm curious if you found this necessary after doing the `.average` calculation, or you tried this first...? What I'm really asking is, does the `.average` calculation return something other than 0 when actors are empty?

File: app/views/movies/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<h1><%= @movie.title %></h1>
+<p>Creation Year: <%= @movie.creation_year %></p>
+<p>Genre: <%= @movie.genre %></p>
+<p>Actors: <% @movie.sort_actors.each do |actor| %></p>
+  <li><%= actor.name %>, <%= actor.age %></li>
+<% end %>
+
+<p>Average Age of Actors: <%= @movie.average_age_of_actors %></p>
+
+<%= form_with url: "/movies/#{@movie.id}/actors", method: :post, local: true do |form| %>
```
Comment (cjsim89) [inline app/views/movies/show.html.erb:10]: Great RESTful request being made here! 👍

File: config/routes.rb [hunk 4]
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  get "/studios", to: "studios#index"
+  get "/movies/:id", to: "movies#show"
+  post "/movies/:id/actors", to: "movie_actors#create"
+  get "/studios/:id", to: "studios#show"
```
Comment (cjsim89) [inline config/routes.rb:7]: Your routes all look RESTful, nice job

File: spec/features/movie_show_page.rb [hunk 5]
```diff
@@ -0,0 +1,119 @@
+require 'rails_helper'
+
+RSpec.describe 'Movies' do
+  describe 'when I visit a movie show page' do
+    it 'shows the movie title, creation year, and genre' do
+      universal = Studio.create!(name: "Universal Studios", location: "Hollywood")
+      jurassic_park = universal.movies.create!(title: "Jurassic Park", creation_year: "1993", genre: "Dinosaurs")
```
Comment (cjsim89) [inline spec/features/movie_show_page.rb:7]: Looks good, though I would try to create more sample data at the start, even if you don't use it, so you have something to test _against_, not just for.

File: spec/features/movie_show_page.rb [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+      MovieActor.create!(movie_id: jurassic_park.id, actor_id: jeff.id)
+      MovieActor.create!(movie_id: jurassic_park.id, actor_id: laura.id)
+      MovieActor.create!(movie_id: jurassic_park.id, actor_id: sam.id)
+
+      visit "/movies/#{jurassic_park.id}"
+
+      expect(laura.name).to appear_before(jeff.name)
+      expect(jeff.name).to appear_before(sam.name)
+      expect(laura.name).to appear_before(sam.name)
```
Comment (cjsim89) [inline spec/features/movie_show_page.rb:32]: Good use of the `orderly` gem

File: spec/features/movie_show_page.rb [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        jurassic_park = universal.movies.create!(title: "Jurassic Park", creation_year: "1993", genre: "Dinosaurs")
+  
+        jeff = Actor.create!(name: "Jeff Goldblum", age: 70)
+        laura = Actor.create!(name: "Laura Dern", age: 55)
+        sam = Actor.create!(name: "Sam Neill", age: 75)
+        samuel = Actor.create!(name: "Samuel L. Jackson", age: 74)
+
+        MovieActor.create!(movie_id: jurassic_park.id, actor_id: jeff.id)
+        MovieActor.create!(movie_id: jurassic_park.id, actor_id: laura.id)
+        MovieActor.create!(movie_id: jurassic_park.id, actor_id: sam.id)
```
Comment (cjsim89) [inline spec/features/movie_show_page.rb:105]: Seeing some repeated code in these 3 tests which is fine, but you may want to consider using `before :each` and setting up instance variables to use throughout your tests, might make it a bit easier next time. Looks good though!

File: spec/features/movie_show_page.rb [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  
+        visit "/movies/#{jurassic_park.id}"
+
+        expect(page).to_not have_content(samuel.name)
```
Comment (cjsim89) [inline spec/features/movie_show_page.rb:109]: Really good test setup here, testing for the negative which then gives weight to the positive below. Keep this up

File: spec/features/studio_index_spec.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+      within "#studio_#{universal.id}" do
+        expect(page).to have_content(harry_potter.title)
+        expect(page).to have_content(harry_potter.creation_year)
+        expect(page).to have_content(harry_potter.genre)
+        expect(page).to have_content(jurassic_park.title)
+        expect(page).to have_content(jurassic_park.creation_year)
+        expect(page).to have_content(jurassic_park.genre)
+        expect(page).to_not have_content(bambi.title)
+        expect(page).to_not have_content(coco.title)
```
Comment (cjsim89) [inline spec/features/studio_index_spec.rb:42]: Great use of both within and to_not! Looks like an accurate & precise test setup to me

File: spec/features/studio_show_page.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    it 'shows the studio name and location' do
+      universal = Studio.create!(name: "Universal Studios", location: "Hollywood")
+
+      visit "/studios/#{universal.id}"
+
+      expect(page).to have_content(universal.name)
+      expect(page).to have_content("Studio Location: #{universal.location}")
+    end
+
+    xit "shows a unique list of all actors that have worked on any of the studio's movies" do
```
Comment (cjsim89) [inline spec/features/studio_show_page.rb:14]: Good start on this; keep going!

File: spec/models/movie_spec.rb [hunk 11]
```diff
@@ -3,5 +3,43 @@
...
+
+      jeff = Actor.create!(name: "Jeff Goldblum", age: 70)
+      laura = Actor.create!(name: "Laura Dern", age: 55)
+      sam = Actor.create!(name: "Sam Neill", age: 75)
+
+      MovieActor.create!(movie_id: jurassic_park.id, actor_id: jeff.id)
+      MovieActor.create!(movie_id: jurassic_park.id, actor_id: laura.id)
+      MovieActor.create!(movie_id: jurassic_park.id, actor_id: sam.id)
+
+      expect(jurassic_park.sort_actors).to eq([laura, jeff, sam])
```
Comment (cjsim89) [inline spec/models/movie_spec.rb:23]: I really recommend having 3 expect lines per model test - one for the basic values like you have here, another testing against _different_ values, then another testing for a 0 value (or, nil/null/[]). I would call that a robust test, whereas this is really just doing the minimal amount to pass.
